src/utils/MapManager.ts

```diff
@@ -0,0 +1,163 @@
+import { convertSpritesheetToIndexArray } from "@shared/canvas/CanvasUtil";
+import { Map } from "src/types/MapType";
+
+export class MapManager {
+  private _map: Map;
+
+  private tileIndexMap: number[][];
+  private spritePixelArray: Uint8Array;
+  private mapPixelArray: Uint8Array;
+
+  constructor(map: Map) {
+    if (map.width <= 0 || map.height <= 0) {
+      throw new Error("Map _width and height must be greater than 0");
+    }
+
+    if (map.mapData.length !== map.width * map.height * map.stride) {
+      throw new Error(`Map data length (${map.mapData.length}) does not match expected size (${map.width * map.height * map.stride})`);
+    }
+
+    this._map = map;
+
+    //FIXME should aready be converted in Project
+    this.spritePixelArray = convertSpritesheetToIndexArray(map.spriteSheet);
+    this.tileIndexMap = this.parseMapData();
+    this.mapPixelArray = this.buildMapPixelArray();
+  }
+
+  public getMapPixelArray(): Uint8Array {
+    return this.mapPixelArray;
+  }
+
+  public getTileAt(x: number, y: number): number | null {
+    if (x < 0 || x >= this._map.width || y < 0 || y >= this._map.height) {
+      return null;
+    }
+    return this.tileIndexMap[y][x];
+  }
+
+  public setTileAt(x: number, y: number, spriteIndex: number): void {
+    if (x < 0 || x >= this._map.width || y < 0 || y >= this._map.height) {
+      throw new Error(`Tile position (${x}, ${y}) is out of bounds`);
+    }
+
+    this.tileIndexMap[y][x] = spriteIndex;
+
+  }
+
+  public getMapDimensions(): { width: number; height: number } {
+    return {
+      width: this._map.width,
+      height: this._map.height
+    };
+  }
+
+  public exportMapData(): string {
+    let result = "";
+    for (let y = 0; y < this._map.height; y++) {
+      for (let x = 0; x < this._map.width; x++) {
+        const tileIndex = this.tileIndexMap[y][x];
+        result += tileIndex.toString(16).padStart(2, "0");
+      }
+    }
+    return result;
+  }
+
+  public updateMapData(new_mapData: string): void {
+    if (new_mapData.length !== this._map.width * this._map.height * 2) {
+      throw new Error(`New map data length (${new_mapData.length}) does not match expected size (${this._map.width * this._map.height * 2})`);
+    }
+
+    this._map.mapData = new_mapData;
+    this.tileIndexMap = this.parseMapData();
+
+    this.mapPixelArray = this.buildMapPixelArray();
+  }
+
+  // PRIVATE
+
+  private parseMapData(): number[][] {
+    const tiles: number[][] = [];
```

Comment:
Please preallocate your array
[https://gamealchemist.wordpress.com/2013/05/01/lets-get-those-javascript-arrays-to-work-fast/](https://gamealchemist.wordpress.com/2013/05/01/lets-get-those-javascript-arrays-to-work-fast/) --> 6th section: "6. pre-allocate your arrays whenever you can."